### PR TITLE
Sync the active versions of the OS and Kernel on boot

### PR DESCRIPTION
**Pull request by mvo5**
cmd/snappy, snappy:fix failover handling of the "active" kernel/os snap

The fallover code may need to fallback on to the previous
OS/Kernel on boot. When this happens we also need to update
the "current" active version of the snap on disk. I.e. when
kernel-v2 fails to boot we want to revert the active kernel
in snappy list to the previous version.

### CODE REVIEW

File: snappy/kernel.go [hunk 1]
```diff
@@ -177,3 +177,39 @@ func kernelOrOsRebootRequired(s *SnapPart) bool {
...
 	return false
 }
+
+func nameAndVersionFromSnap(snap string) (string, string) {
+	name := strings.Split(snap, "_")[0]
+	ver := strings.Split(snap, "_")[1]
+	return name, strings.Split(ver, ".snap")[0]
+}
+
+// SyncBoot synchronizes the active kernel and OS snap version with
```
Comment (elopio) [inline snappy/kernel.go]: s/version/versions

File: snappy/kernel.go [hunk 2]
```diff
@@ -177,3 +177,39 @@ func kernelOrOsRebootRequired(s *SnapPart) bool {
...
+// the version that actually booted. The reason this is needed is
```
Comment (elopio) [inline snappy/kernel.go]: s/version/versions

Comment (elopio) [inline snappy/kernel.go]: s/The reason this is needed is/This is needed

File: snappy/kernel.go [hunk 3]
```diff
@@ -177,3 +177,39 @@ func kernelOrOsRebootRequired(s *SnapPart) bool {
...
+// because a system may install "os=v2" but that fails to boot. The
+// bootloader fallback logic will revert to "os=v1" but on the
+// filesystem snappy still has the "active" version set to "v2" which
+// is misleading. This code will check what kernel/os booted and set
+// those versions active.
+func SyncBoot() error {
+	kernelSnap, _ := getBootVar("snappy_kernel")
+	osSnap, _ := getBootVar("snappy_os")
```
Comment (elopio) [inline snappy/kernel.go]: isn't it weird that we are handling the os snap in the kernel.go file?

**Comment by elopio** [comment]
lgtm. I just left up there a naive question. And here comes another one:
How are the integration tests passing if this is still failing?

**Comment by elopio** [comment]
Ah, nevermind, I found the skip just after clicking send, and this bug: https://bugs.launchpad.net/snappy/+bug/1534029
Can you please remove the skip from failover_test?

**Comment by mvo5** [comment]
@elopio Thanks for the review! I fixed your review comments. I did not yet re-enable the integration tests, just verified this locally. This is because the "snappy booted" command will use the /usr/bin/snappy which is still the old version when the integration tests run. So we can re-enable the integration tests once this code has landed and a new ubuntu-core snap is build and available for the integration tests.

File: snappy/kernel_os.go [hunk 4]
```diff
@@ -177,3 +177,39 @@ func kernelOrOsRebootRequired(s *SnapPart) bool {
...
 }
+
+func nameAndVersionFromSnap(snap string) (string, string) {
+	name := strings.Split(snap, "_")[0]
+	ver := strings.Split(snap, "_")[1]
+	return name, strings.Split(ver, ".snap")[0]
+}
+
+// SyncBoot synchronizes the active kernel and OS snap versions with
+// the versions that actually booted. This is needed is because a
```
Comment (elopio) [inline snappy/kernel_os.go]: Remove the second "is".

**Comment by elopio** [comment]
@mvo5 understood about the integration test.
+1, with one typo left to fix.

**Comment by mvo5** [comment]
retest this please

**Comment by chipaca** [comment]
LGTM. Not merging myself because the commit message is not up to it.

**Comment by chipaca** [comment]
Wait, yes it is.

**Pull request merged**